src/common/libflux/future.c

```diff
@@ -0,0 +1,471 @@
+        errno = ENOMEM;
+        goto error;
+    }
+    then->r = r;
+    if (!(then->prepare = flux_prepare_watcher_create (r, prepare_cb, arg)))
+        goto error;
+    if (!(then->check = flux_check_watcher_create (r, check_cb, arg)))
+        goto error;
+    if (!(then->idle = flux_idle_watcher_create (r, NULL, NULL)))
+        goto error;
+    flux_watcher_start (then->prepare);
+    flux_watcher_start (then->check);
+    return then;
+error:
+    then_context_destroy (then);
+    return NULL;
+}
+
+static int then_context_set_timeout (struct then_context *then,
+                                     double timeout, void *arg)
+{
+    if (then) {
+        if (!then->timer) {
+            then->timer = flux_timer_watcher_create (then->r, timeout, 0.,
+                                                     timer_cb, arg);
+            if (!then->timer)
+                return -1;
+        }
+        else {
+            flux_timer_watcher_reset (then->timer, timeout, 0.);
+        }
+        flux_watcher_start (then->timer);
+    }
+    return 0;
+}
+
+static void then_context_clear_timer (struct then_context *then)
+{
+    if (then)
+        flux_watcher_stop (then->timer);
+}
+
+/* Destroy a future.
+ */
+void flux_future_destroy (flux_future_t *f)
+{
+    int saved_errno = errno;
+    if (f) {
+        zhash_destroy (&f->aux);
+        if (f->result) {
+            if (f->result_free)
+                f->result_free (f->result);
+        }
+        now_context_destroy (f->now);
+        then_context_destroy (f->then);
+        free (f);
+    }
+    errno = saved_errno;
+}
+
+/* Create a future.
+ */
+flux_future_t *flux_future_create (flux_reactor_t *r)
+{
+    flux_future_t *f = calloc (1, sizeof (*f));
+    if (!f) {
+        errno = ENOMEM;
+        goto error;
+    }
+    if (!r) {
+        errno = EINVAL;
+        goto error;
+    }
+    f->r = r;
+    return f;
+error:
+    flux_future_destroy (f);
+    return NULL;
+}
+
+/* Set/update timeout
+ * If timeout expires, result will be an error with errno == ETIMEDOUT
+ */
+int flux_future_set_timeout (flux_future_t *f, double timeout)
+{
+    f->timeout = timeout;
+    f->timeout_valid = 1;
+    if (now_context_set_timeout (f->now, timeout, f) < 0)
+        return -1;
+    if (then_context_set_timeout (f->then, timeout, f) < 0)
+        return -1;
+    return 0;
+}
+
+void flux_future_set_flux (flux_future_t *f, flux_t *h)
+{
+    f->h = h;
+}
+
+flux_t *flux_future_get_flux (flux_future_t *f)
+{
+    flux_t *h = NULL;
+
+    if (!f->h) {
+        errno = EINVAL;
+        goto done;
+    }
+    if (!f->now || !f->now->running) {
+        h = f->h;
+    }
+    else {
+        if (!f->now->h) {
+            if (!(f->now->h = flux_clone (f->h)))
+                goto done;
+            if (flux_set_reactor (f->now->h, f->now->r) < 0)
+                goto done;
+        }
+        h = f->now->h;
+    }
+done:
+    return h;
+}
+
+static void run_now (flux_future_t *f, int reactor_flags)
```

Comment:
At the very least, this function needs a good code comment.  A better name might be in order too.